conductor/src/action.hpp

```diff
@@ -0,0 +1,192 @@
+
+#ifndef COCAINE_CONDUCTOR_ACTION_HPP
+#define COCAINE_CONDUCTOR_ACTION_HPP
+
+
+namespace cocaine { namespace isolate { namespace conductor {
+
+namespace action {
+
+struct cancellation_t:
+    public api::cancellation_t
+{
+    typedef std::shared_ptr<action_t> parent_ptr;
+    typedef std::weak_ptr<action_t> parent_wptr;
+
+    parent_wptr m_parent;
+
+    cancellation_t(parent_ptr parent);
+
+    virtual void cancel();
+};
+
+
+class action_t:
+    public enable_shared_from_this<action_t>
+{
+    // enum states {
+    //     st_pending = 0,
+    //     st_done,
+    //     st_cancelled
+    // };
+
+protected:
+
+    uint64_t             m_request_id;
+    shared_ptr<client_t> m_parent;
+
+    shared_ptr<io::basic_dispatch_t> m_dispatch;
+
+public:
+
+    struct states {
+        const static int st_pending = 0;
+        const static int st_done = 1;
+        const static int st_cancelled = 2;
+    };
+
+    std::atomic<int>     m_state;
+
+
+    typedef session<asio::ip::tcp> session_type;
+
+    action_t(shared_ptr<client_t> client):
+        m_state(states::st_pending),
+        m_request_id(++client->max_request_id),
+        m_parent(client)
+    {}
+
+    virtual
+    shared_ptr<io::basic_dispatch_t>
+    dispatch() = 0;
+
+    virtual void send(shared_ptr<session_type> session) = 0;
+
+    virtual void on_done(const dynamic_t& result) = 0;
+
+    virtual void on_error(const std::error_code& ec) = 0;
+
+    uint64_t
+    id(){
+        return m_request_id;
+    }
+
+    virtual
+    auto cancellation() -> std::unique_ptr<api::cancellation_t> {
+        std::unique_ptr<api::cancellation_t> c(new cancellation_t(shared_from_this()));
+        return std::move(c);
+    }
+
+    virtual
+    void
+    run(){
+        m_parent->enqueue(shared_from_this());
+    }
+
+    virtual
+    void
+    reset(){
+        m_dispatch.reset();
+    }
+
+    virtual
+    void
+    cancel(){
+        COCAINE_LOG_DEBUG(m_parent->m_log, "request[%d]: cancelling", m_request_id);
+        int expected = states::st_pending;
```

Comment:
there was some compiler error about it not able to get `int&` from `states::st_pending`